binary_search_tree/tree.py

```diff
@@ -7,53 +7,116 @@ def __init__(self, key, val = None):
         self.right = None
+    
+    def dict(self):
+        return {"key": self.key, "value": self.value}
         
 
 
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add(self, key, value = None):
-        pass
+        tree = TreeNode(key, value)
+
+        if self.root == None:
+            self.root = tree
+        else:
+            current = self.root
+            
+            while current != None:
+                if key == current.key:
+                    return None
+                if key < current.key:
+                    if current.left == None:
+                        current.left = tree
+                    current = current.left
+                elif key > current.key:
+                    if current.right == None:
+                        current.right = tree
+                    current = current.right
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+        current = self.root
+        while current != None:
+            if key == current.key:
+                return current.value
+            elif key > current.key:
+                current = current.right
+            else:
+                current = current.left 
+        return None
+
+    def inorder_helper(self, current, traversal_list):
+        if current == None:
+            return
+
+        self.inorder_helper(current.left, traversal_list)
+        traversal_list.append(current.dict())
+        self.inorder_helper(current.right, traversal_list)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def inorder(self):
-        pass
+        traversal_list = []
+        self.inorder_helper(self.root, traversal_list)
+        return traversal_list   
+
+    def preorder_helper(self, current, traversal_list):
+        if current == None:
+            return
 
-    # Time Complexity: 
-    # Space Complexity:     
+        traversal_list.append(current.dict())
+        self.preorder_helper(current.left, traversal_list)
+        self.preorder_helper(current.right, traversal_list)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)    
     def preorder(self):
-        pass
+        traversal_list = []
+        self.preorder_helper(self.root, traversal_list)
+        return traversal_list
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def postorder_helper(self, current, traversal_list):
+        if current == None:
+            return
+
+        self.postorder_helper(current.left, traversal_list)
+        self.postorder_helper(current.right, traversal_list)
+        traversal_list.append(current.dict())
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def postorder(self):
```

Comment:
👍, however since you are building a list of all tree node values the space complexity is O(n)